OS to EE Rolling Restarts Doc Updates (#1145)
hazelcast/hz-docs: docs/modules/migrate/pages/rolling-restart.adoc

= Upgrade Cluster to {enterprise-product-name}

> :description: You can use an in-place rolling restart to move each member on your Open Source cluster to an Enterprise cluster.

Reviewer: This wording is a bit weird. 

Maybe something like
```suggestion
:description: You can use an in-place rolling restart to change each member to upgrade your cluster from Open Source to Enterprise cluster.
```

Author: changed to:

You can upgrade your Open Source cluster to an Enterprise cluster using an in-place rolling restart to change each member in turn.

Reviewer: Good.

{description}

This approach ensures that you can move your cluster without interruption to the service. 
Similar to the rolling upgrade process, a rolling restart allows you to move one member at a time.

The codebase of your {open-source-product-name} cluster must match the target {enterprise-product-name} release. 

> Currently, rolling restart can only be used to update your 5.5 {open-source-product-name} cluster members.

Reviewer: You can replace 5.5 with current version.

Going forwards, it will be possible to update each version from OS to EE.

Author: I have removed the line, as I think it is covered by the previous statement on matching codebases.

The issue is - if we change to current version - this could be interpreted as their OS current version, so any version, which would be incorrect.

Feel free to disagree ;-)

== Best Practice

Before starting a rolling restart, consider the following best practices:

> * Backup your IMap and ReplicatedMap data structures. This is not essential, but it is recommended practice

Reviewer: How should they do the backup? AFAIK we don't provide any tooling for that. (DMT can transfer data between clusters, but not simply export...)

Author: I have added "using your usual method", but if this is too much of a fudge (or not something that is generally/easily done) I'd suggest removing this bullet from the list.

Reviewer: I suppose that's good.

> * If the {open-source-product-name} member is a different version to the target {enterprise-product-name} member, update the codebase to match the target {enterprise-product-name} release

Reviewer: Wouldn't this defeat the whole purpose of the OS to EE restart?

Say (in the future) user is running 5.5.x OS cluster, and they want to upgrade to 5.6.x EE.

This says to update OS cluster to 5.5 first to 5.6, and then do OS to EE restart.
But they can't update the cluster without losing the data.

What they need to do is OS to EE restart first on 5.5 and the do RU from 5.5 to 5.6.

Author: Possibly my misunderstanding (and related to the comment above over 5.5 v current version, I understood that the OS members need to be 5.5 before they can be moved.

If I was incorrect, this bullet can be removed. Otherwise, we need to clarify somehow that the OS must be on 5.5; and that statement would need to be removed for the next release.

+

> **[RJL: Do we need to include any additional information to cover moving from - for example - 5.3 with CP Subsystem to Enterprise 5.5?]**

Reviewer: This was removed from the scope, currently there is no good solution for that.

Author: Not ideal. And I'm not sure we can manage that expectation through the docs without it - removed question; but I'd note there are likely to be support calls around this if we cannot provide the solution 😬

* Ensure that the version of JDK meets the minimum required for the target {enterprise-product-name} release
* Plan sufficient time for the rolling restart to complete, and remember that the member cannot change during the move

== Perform a Rolling Restart

To move your {open-source-product-name} cluster to {enterprise-product-name}, complete the following steps:

. Gracefully shut down an existing {open-source-product-name} member.
+
See xref:maintain-cluster:shutdown.adoc#shutting-down-a-hazelcast-member[Shutting Down a Hazelcast Member].

. Wait for all partitions to migrate to the rest of the cluster.
+
During migrations, members cannot join or leave the cluster.

. Upgrade the member's codebase to match the {enterprise-product-name} version.

. Start the member and wait until it joins the {enterprise-product-name} cluster.

. Repeat this process for each member in the {open-source-product-name} cluster.

> . When all members on the {open-source-product-name} cluster have been moved to the {enterprise-product-name} cluster, enable and configure the 

Reviewer: I would link dynamic configuration here. It's the only well supported way to enable EE features on a running cluster.
